### PR TITLE
Refactor fit bragg peaks and add unit tests

### PR DESCRIPTION
**Pull request by MialLewis**
**Description of work:**

Refactors the `_fit_bragg_peaks` portion of `EVSCalibrationFit`, making it easier to follow and unit test. This was probably the most convoluted function.

Unit tests have been added where the code has been refactored.

**To test:**
Review code.
Review test changes.
Ensure all automated tests pass.
Ensure the tests in `test_system.py` pass locally (You can run the system tests using `python -m unittest discover -s ./unpackaged/vesuvio_calibration/tests/system` from the root of the repository.) Note that `test_fit_bragg_peaks_copper` and `test_fit_bragg_peaks_lead` are unreliable and fail on different detectors on different PCs. Please post the stack trace if they fail on your machine.

### CODE REVIEW

File: unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py [hunk 1]
```diff
@@ -379,114 +379,155 @@ def _normalise_sample_by_background(self):
...
-    CreateEmptyTableWorkspace(OutputWorkspace=peaks_table)
+    output_parameters_tbl_name = self._output_workspace_name + '_Peak_Parameters'
+    self._create_output_parameters_table_ws(output_parameters_tbl_name, num_estimated_peaks)
 
-    #PARAM NAMES FOR LINEAR BACKGROUND
-    param_names = ['f0.A0', 'f0.A1']
+    output_workspaces = []
+    for index, estimated_peak_positions in enumerate(estimated_peak_positions_all_spec.transpose()):
+        spec_number = self._spec_list[0]+index
+        self._prog_reporter.report("Fitting to spectrum %d" % spec_number)
```
Comment (robertapplin) [inline unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py:393]: Is it possible to modernise this file to use f-strings?

```suggestion
        self._prog_reporter.report(f"Fitting to spectrum {spec_number}")
```

File: unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py [hunk 2]
```diff
@@ -379,114 +379,155 @@ def _normalise_sample_by_background(self):
...
+        find_peaks_output_name = self._sample + '_peaks_table_%d' % spec_number
+        fit_peaks_output_name = self._output_workspace_name + '_Spec_%d' % spec_number
+        fit_results = self._fit_peaks_to_spectra(index, spec_number, estimated_peak_positions, find_peaks_output_name,
+                                                 fit_peaks_output_name)
 
-    #GENERATE ERROR NAMES FROM PARAM NAMES, ZIP TO GET COL NAMES
-    err_names = [name + '_Err' for name in param_names]
-    col_names = [element for tupl in zip(param_names, err_names) for element in tupl]
+        fit_results_u = None
+        if not fit_results['status'] == "success":
```
Comment (robertapplin) [inline unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py:401]:
```suggestion
        if fit_results['status'] != "success":
```

File: unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py [hunk 3]
```diff
@@ -379,114 +379,155 @@ def _normalise_sample_by_background(self):
...
+            self._prog_reporter.report("Fitting to spectrum %d without constraining parameters" % spec_number)
+            fit_results_u = self._fit_peaks_to_spectra(index, spec_number, estimated_peak_positions, find_peaks_output_name,
```
Comment (robertapplin) [inline unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py:403]: This is just a question for my own understanding, what does the `_u` denote at the end of this variable?

File: unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py [hunk 4]
```diff
@@ -379,114 +379,155 @@ def _normalise_sample_by_background(self):
...
+      row = [element for tupl in zip(row_values, row_errors) for element in tupl]
+
+      mtd[output_table_name].addRow([spec_num] + row)
+
+  def _get_output_and_clean_workspaces(self, unconstrained_fit_performed, unconstrained_fit_selected, find_peaks_output_name,
+                                       fit_peaks_output_name):
+      find_peaks_output_name_u = self._get_unconstrained_ws_name(find_peaks_output_name)
+      fit_peaks_output_name_u = self._get_unconstrained_ws_name(fit_peaks_output_name)
+
+      DeleteWorkspace(fit_peaks_output_name + '_NormalisedCovarianceMatrix')
```
Comment (robertapplin) [inline unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py:515]: The `_Workspace` and other suffixes seemed to be used multiple times here. Would it be possible to make it a named constant at the top of the file e.g. `WORKSPACE_SUFFIX = "_Workspace"`?

**Review by robertapplin — CHANGES_REQUESTED** [review]
Lots of new tests which is great 👍 I have a few code suggestions but feel free to push back if you disagree

File: unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py (hunk 1 above)
Comment (MialLewis) [reply unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py:393]: I need to go through the whole script and update a few things, use of `f-strings` is one of them. I will do this in a subsequent PR once this batch has been merged.

File: unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py (hunk 3 above)
Comment (MialLewis) [reply unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py:403]: It stands for `unconstrained`, I don't see any reason not to change the variable name to make this clearer.

File: unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py (hunk 4 above)
Comment (MialLewis) [reply unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py:516]: I think it would be good to rationalise workspace naming/the use of suffixes throughout the whole script, as it's quite hard to follow. I'd prefer to do this at once in a subsequent PR though.

File: unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py [hunk 5]
```diff
@@ -379,114 +379,155 @@ def _normalise_sample_by_background(self):
...
-        else:
-            output_workspaces.append(fit_output_name + '_Workspace')
+  def _create_output_parameters_table_ws(self, output_table_name, num_estimated_peaks):
+    CreateEmptyTableWorkspace(OutputWorkspace=output_table_name)
+
+    col_headers = self._generate_column_headers(num_estimated_peaks)
+
+    mtd[output_table_name].addColumn('int', 'Spectrum')
+    for name in col_headers:
+        mtd[output_table_name].addColumn('double', name)
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply unpackaged/vesuvio_calibration/calibrate_vesuvio_uranium_martyn_MK5.py]: Good suggestion, I've made this change.

**Comment by robertapplin** [comment]
I get one failure locally:
```
FAIL: test_fit_bragg_peaks_copper (test_system.TestEVSCalibrationFit)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\mlc47243\AppData\Local\mambaforge\envs\mantidnightly\lib\site-packages\mock\mock.py", line 1417, in patched
    return func(*newargs, **newkeywargs)
  File "C:\Users\mlc47243\Documents\mantid-dev\vesuvio\unpackaged\vesuvio_calibration\tests\system\test_system.py", line 374, in test_fit_bragg_peaks_copper
    self.assert_fitted_positions_match_expected(expected_values, params_table,
  File "C:\Users\mlc47243\Documents\mantid-dev\vesuvio\unpackaged\vesuvio_calibration\tests\system\test_system.py", line 474, in assert_fitted_positions_match_expected
    raise AssertionError(error_dict)
AssertionError: {'f3.LorentzPos': ['Element 15: Expected 11323.436309381943, found 4574.390063160628. atol 6749.0462462213145, rtol 1.4753980646674738,max tol: 0.1']}
```

is this expected?

**Comment by MialLewis** [comment]
Yes thanks expected, going to let that failure go for now and fix in a later PR.

**Review by robertapplin — APPROVED** [review]
Thanks, happy to approve in that case 👍
